src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultBoundAttributeDescriptorBuilder.cs

```diff
@@ -94,26 +87,19 @@ public override void BindAttributeParameter(Action<BoundAttributeParameterDescri
 
     public BoundAttributeDescriptor Build()
     {
-        var diagnostics = Validate();
-        if (_diagnostics != null)
-        {
-            diagnostics ??= new();
-            diagnostics.UnionWith(_diagnostics);
-        }
+        using var _ = HashSetPool<RazorDiagnostic>.GetPooledObject(out var diagnostics);
 
-        var parameters = Array.Empty<BoundAttributeParameterDescriptor>();
-        if (_attributeParameterBuilders != null)
-        {
-            // Attribute parameters are case-sensitive.
-            var parameterset = new HashSet<BoundAttributeParameterDescriptor>(BoundAttributeParameterDescriptorComparer.Default);
-            for (var i = 0; i < _attributeParameterBuilders.Count; i++)
-            {
-                parameterset.Add(_attributeParameterBuilders[i].Build());
-            }
+        Validate(diagnostics);
 
-            parameters = parameterset.ToArray();
+        if (_diagnostics is { } existingDiagnostics)
+        {
+            diagnostics.UnionWith(existingDiagnostics);
         }
 
+        var parameters = _attributeParameterBuilders is { } attributeParameterBuilders
```

Comment:
```suggestion
        var parameters = _attributeParameterBuilders is not null attributeParameterBuilders
```

Reply:
Sorry that this doesn't work today! :smile:

Reply:
I'm updating this with an extension method to avoid duplicated code. So, you won't see this pattern matching anymore. :smile: